The job **`x86_64-gnu-llvm-14`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/4971890826/jobs/8896728327) [(plain)](https://github.com/rust-lang/rust/commit/00b8c0eea3fa1b6c6c8bb397b67a52d6aa947912/checks/13464338347/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
To only update this specific test, also pass `--test-args panics/default-backtrace-ice.rs`

error: 1 errors occurred comparing output.
status: exit status: 101
command: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/panics/default-backtrace-ice.rs" "-Zthreads=1" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Zdeduplicate-diagnostics=no" "-Cstrip=debuginfo" "--remap-path-prefix=/checkout/tests/ui=fake-test-src-base" "--emit" "metadata" "-C" "prefer-dynamic" "--out-dir" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/panics/default-backtrace-ice" "-A" "unused" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/panics/default-backtrace-ice/auxiliary" "-Z" "treat-err-as-bug=1"
stdout: none
error[E0425]: cannot find value `missing_ident` in this scope
  --> fake-test-src-base/panics/default-backtrace-ice.rs:21:13
   |
   |
LL | fn main() { missing_ident; }


thread 'rustc' panicked at 'aborting due to `-Z treat-err-as-bug=1`', compiler/rustc_errors/src/lib.rs:1711:30
   0:     0x7ff4b84e3494 - <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt::h7cb2601110671950
   1:     0x7ff4b854a048 - core::fmt::write::he95f93868e144b78
   2:     0x7ff4b84d7c41 - std::io::Write::write_fmt::h52122b99cb8b71d6
   3:     0x7ff4b84e32a1 - std::sys_common::backtrace::print::h021a40a15153f1ce
   3:     0x7ff4b84e32a1 - std::sys_common::backtrace::print::h021a40a15153f1ce
   4:     0x7ff4b84e63fa - std::panicking::default_hook::{{closure}}::ha19f26333d1f985e
   5:     0x7ff4b84e60dc - std::panicking::default_hook::h997763420b2dc24e
   6:     0x7ff4b8faa8bb - rustc_driver_impl[ce86b870382ee6da]::install_ice_hook::{closure#0}
   7:     0x7ff4b84e6b07 - std::panicking::rust_panic_with_hook::h6b19f42d65e969e0
   8:     0x7ff4b84e6849 - std::panicking::begin_panic_handler::{{closure}}::ha5e419268b20295d
   9:     0x7ff4b84e3956 - std::sys_common::backtrace::__rust_end_short_backtrace::h0b7c3b110996144b
  11:     0x7ff4b849b0d3 - core::panicking::panic_fmt::hca81948884174e02
  11:     0x7ff4b849b0d3 - core::panicking::panic_fmt::hca81948884174e02
  12:     0x7ff4bbee77f7 - <rustc_errors[30ad7a69ca27aec9]::HandlerInner>::panic_if_treat_err_as_bug
  13:     0x7ff4bbee6c1c - <rustc_errors[30ad7a69ca27aec9]::HandlerInner>::emit_diagnostic::{closure#2}
  14:     0x7ff4b90ce64f - rustc_interface[1be889d4954d160]::callbacks::track_diagnostic
  15:     0x7ff4bbee63d7 - <rustc_errors[30ad7a69ca27aec9]::HandlerInner>::emit_diagnostic
  16:     0x7ff4bbee53de - <rustc_errors[30ad7a69ca27aec9]::Handler>::emit_diagnostic
  17:     0x7ff4bbeef69a - <rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed as rustc_errors[30ad7a69ca27aec9]::diagnostic_builder::EmissionGuarantee>::diagnostic_builder_emit_producing_guarantee
  18:     0x7ff4ba07edab - <rustc_resolve[f957c7b68148594b]::Resolver>::report_errors
  19:     0x7ff4ba05997e - <rustc_session[90db9b8c480f4336]::session::Session>::time::<(), <rustc_resolve[f957c7b68148594b]::Resolver>::resolve_crate::{closure#0}>
  20:     0x7ff4ba0aa36d - <rustc_resolve[f957c7b68148594b]::Resolver>::resolve_crate
  21:     0x7ff4b9089c1d - rustc_interface[1be889d4954d160]::passes::resolver_for_lowering
  22:     0x7ff4ba9ad83a - rustc_query_impl[20460c98a3456e2]::plumbing::__rust_begin_short_backtrace::<<rustc_query_impl[20460c98a3456e2]::queries::resolver_for_lowering as rustc_query_system[3992e411d3f662b7]::query::config::QueryConfig<rustc_query_impl[20460c98a3456e2]::plumbing::QueryCtxt>>::compute::{closure#0}, rustc_middle[9ad29ca9d0d612af]::query::erase::Erased<[u8; 8usize]>>
  23:     0x7ff4baccc713 - rustc_query_system[3992e411d3f662b7]::query::plumbing::try_execute_query::<rustc_query_impl[20460c98a3456e2]::queries::resolver_for_lowering, rustc_query_impl[20460c98a3456e2]::plumbing::QueryCtxt>
  24:     0x7ff4baa6b763 - rustc_query_impl[20460c98a3456e2]::get_query::resolver_for_lowering::__rust_end_short_backtrace
  25:     0x7ff4b8fc6abd - <rustc_middle[9ad29ca9d0d612af]::ty::context::GlobalCtxt>::enter::<rustc_driver_impl[ce86b870382ee6da]::run_compiler::{closure#1}::{closure#2}::{closure#2}, &rustc_data_structures[a53546a1792cd81a]::steal::Steal<(rustc_middle[9ad29ca9d0d612af]::ty::ResolverAstLowering, alloc[9ba026c34f0ac746]::rc::Rc<rustc_ast[dc4c8f671980910a]::ast::Crate>)>>
  26:     0x7ff4b8ffcb5f - <rustc_interface[1be889d4954d160]::interface::Compiler>::enter::<rustc_driver_impl[ce86b870382ee6da]::run_compiler::{closure#1}::{closure#2}, core[5ff145dc1dc74fd]::result::Result<core[5ff145dc1dc74fd]::option::Option<rustc_interface[1be889d4954d160]::queries::Linker>, rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>
  27:     0x7ff4b8fc5950 - rustc_span[f5d6d93b9d80788f]::set_source_map::<core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>, rustc_interface[1be889d4954d160]::interface::run_compiler<core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>, rustc_driver_impl[ce86b870382ee6da]::run_compiler::{closure#1}>::{closure#0}::{closure#0}>
  28:     0x7ff4b8fb7a79 - <scoped_tls[3bb03c838efd0f64]::ScopedKey<rustc_span[f5d6d93b9d80788f]::SessionGlobals>>::set::<rustc_interface[1be889d4954d160]::interface::run_compiler<core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>, rustc_driver_impl[ce86b870382ee6da]::run_compiler::{closure#1}>::{closure#0}, core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>
  29:     0x7ff4b8fcfc46 - std[d7ca55dceeeccb01]::sys_common::backtrace::__rust_begin_short_backtrace::<rustc_interface[1be889d4954d160]::util::run_in_thread_pool_with_globals<rustc_interface[1be889d4954d160]::interface::run_compiler<core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>, rustc_driver_impl[ce86b870382ee6da]::run_compiler::{closure#1}>::{closure#0}, core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>
  30:     0x7ff4b8ffe196 - std[d7ca55dceeeccb01]::panicking::try::<core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>, core[5ff145dc1dc74fd]::panic::unwind_safe::AssertUnwindSafe<<std[d7ca55dceeeccb01]::thread::Builder>::spawn_unchecked_<rustc_interface[1be889d4954d160]::util::run_in_thread_pool_with_globals<rustc_interface[1be889d4954d160]::interface::run_compiler<core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>, rustc_driver_impl[ce86b870382ee6da]::run_compiler::{closure#1}>::{closure#0}, core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>::{closure#1}::{closure#0}>>
  31:     0x7ff4b8fbd32d - <<std[d7ca55dceeeccb01]::thread::Builder>::spawn_unchecked_<rustc_interface[1be889d4954d160]::util::run_in_thread_pool_with_globals<rustc_interface[1be889d4954d160]::interface::run_compiler<core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>, rustc_driver_impl[ce86b870382ee6da]::run_compiler::{closure#1}>::{closure#0}, core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>::{closure#0}::{closure#0}, core[5ff145dc1dc74fd]::result::Result<(), rustc_span[f5d6d93b9d80788f]::ErrorGuaranteed>>::{closure#1} as core[5ff145dc1dc74fd]::ops::function::FnOnce<()>>::call_once::{shim:vtable#0}
  32:     0x7ff4b84f343e - std::sys::unix::thread::Thread::new::thread_start::hac3846b6a19ada3d
  33:     0x7ff4b828eb43 - <unknown>
  34:     0x7ff4b8320a00 - <unknown>
  35:                0x0 - <unknown>
error: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/issues/new?labels=C-bug%2C+I-ICE%2C+T-compiler&template=ice.md


note: rustc 1.71.0-nightly (c19f4dcd7 2023-05-14) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -C codegen-units=1 -Z ui-testing -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z deduplicate-diagnostics=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0 -Z treat-err-as-bug=1
query stack during panic:
#0 [resolver_for_lowering] getting the resolver for lowering
end of query stack
------------------------------------------
```

</details>

--- a/compiler/rustc_query_impl/src/plumbing.rs
+++ b/compiler/rustc_query_impl/src/plumbing.rs
@@ -488,6 +488,18 @@ macro_rules! expand_if_cached {
     };
 }
 
+/// Don't show the backtrace for query system by default
+/// use `RUST_BACKTRACE=full` to show all the backtraces
+#[inline(never)]
+pub fn __rust_begin_short_backtrace<F, T>(f: F) -> T
+where
+    F: FnOnce() -> T,
+{
+    let result = f();
+    std::hint::black_box(());
+    result
+}
+
 // NOTE: `$V` isn't used here, but we still need to match on it so it can be passed to other macros
 // invoked by `rustc_query_append`.
 macro_rules! define_queries {
@@ -498,21 +510,25 @@ macro_rules! define_queries {
             use super::*;
 
             $(
-                #[inline(always)]
-                #[tracing::instrument(level = "trace", skip(tcx))]
-                pub(super) fn $name<'tcx>(
-                    tcx: TyCtxt<'tcx>,
-                    span: Span,
-                    key: query_keys::$name<'tcx>,
-                    mode: QueryMode,
-                ) -> Option<Erase<query_values::$name<'tcx>>> {
-                    get_query_incr(
-                        queries::$name::config(tcx),
-                        QueryCtxt::new(tcx),
-                        span,
-                        key,
-                        mode
-                    )
+                // Adding `__rust_end_short_backtrace` marker to backtraces so that we emit the frames
+                // when `RUST_BACKTRACE=1`, add a new mod with `$name` here is to allow duplicate naming
+                pub mod $name {
+                    use super::*;
+                    #[inline(never)]
+                    pub fn __rust_end_short_backtrace<'tcx>(
+                        tcx: TyCtxt<'tcx>,
+                        span: Span,
+                        key: query_keys::$name<'tcx>,
+                        mode: QueryMode,
+                    ) -> Option<Erase<query_values::$name<'tcx>>> {
+                        get_query_incr(
+                            queries::$name::config(tcx),
+                            QueryCtxt::new(tcx),
+                            span,
+                            key,
+                            mode
+                        )
+                    }
                 }
             )*
         }
@@ -521,20 +537,22 @@ macro_rules! define_queries {
             use super::*;
 
             $(
-                #[inline(always)]
-                #[tracing::instrument(level = "trace", skip(tcx))]
-                pub(super) fn $name<'tcx>(
-                    tcx: TyCtxt<'tcx>,
-                    span: Span,
-                    key: query_keys::$name<'tcx>,
-                    __mode: QueryMode,
-                ) -> Option<Erase<query_values::$name<'tcx>>> {
-                    Some(get_query_non_incr(
-                        queries::$name::config(tcx),
-                        QueryCtxt::new(tcx),
-                        span,
-                        key,
-                    ))
+                pub mod $name {
+                    use super::*;
+                    #[inline(never)]
+                    pub fn __rust_end_short_backtrace<'tcx>(
+                        tcx: TyCtxt<'tcx>,
+                        span: Span,
+                        key: query_keys::$name<'tcx>,
+                        __mode: QueryMode,
+                    ) -> Option<Erase<query_values::$name<'tcx>>> {
+                        Some(get_query_non_incr(
+                            queries::$name::config(tcx),
+                            QueryCtxt::new(tcx),
+                            span,
+                            key,
+                        ))
+                    }
                 }
             )*
         }
@@ -542,11 +560,11 @@ macro_rules! define_queries {
         pub(crate) fn engine(incremental: bool) -> QueryEngine {
             if incremental {
                 QueryEngine {
-                    $($name: get_query_incr::$name,)*
+                    $($name: get_query_incr::$name::__rust_end_short_backtrace,)*
                 }
             } else {
                 QueryEngine {
-                    $($name: get_query_non_incr::$name,)*
+                    $($name: get_query_non_incr::$name::__rust_end_short_backtrace,)*
                 }
             }
         }
@@ -578,10 +596,15 @@ macro_rules! define_queries {
                         query_cache: offset_of!(QueryCaches<'tcx> => $name),
                         cache_on_disk: |tcx, key| ::rustc_middle::query::cached::$name(tcx, key),
                         execute_query: |tcx, key| erase(tcx.$name(key)),
-                        compute: |tcx, key| query_provided_to_value::$name(
-                            tcx,
-                            call_provider!([$($modifiers)*][tcx, $name, key])
-                        ),
+                        compute: |tcx, key| {
+                            use crate::plumbing::__rust_begin_short_backtrace;
+                            __rust_begin_short_backtrace(||
+                                query_provided_to_value::$name(
+                                    tcx,
+                                    call_provider!([$($modifiers)*][tcx, $name, key])
+                                )
+                            )
+                        },
                         can_load_from_disk: should_ever_cache_on_disk!([$($modifiers)*] true false),
                         try_load_from_disk: should_ever_cache_on_disk!([$($modifiers)*] {
                             |tcx, key, prev_index, index| {

--- a/library/std/src/sys_common/backtrace.rs
+++ b/library/std/src/sys_common/backtrace.rs
@@ -68,13 +68,17 @@ unsafe fn _print_fmt(fmt: &mut fmt::Formatter<'_>, print_fmt: PrintFmt) -> fmt::
         }
 
         let mut hit = false;
-        let mut stop = false;
         backtrace_rs::resolve_frame_unsynchronized(frame, |symbol| {
             hit = true;
+
+            // Any frames between `__rust_begin_short_backtrace` and `__rust_end_short_backtrace`
+            // are omitted from the backtrace in short mode, `__rust_end_short_backtrace` will be
+            // called before the panic hook, so we won't ignore any frames if there is no
+            // invoke of `__rust_begin_short_backtrace`.
             if print_fmt == PrintFmt::Short {
                 if let Some(sym) = symbol.name().and_then(|s| s.as_str()) {
                     if start && sym.contains("__rust_begin_short_backtrace") {
-                        stop = true;
+                        start = false;
                         return;
                     }
                     if sym.contains("__rust_end_short_backtrace") {
@@ -88,9 +92,6 @@ unsafe fn _print_fmt(fmt: &mut fmt::Formatter<'_>, print_fmt: PrintFmt) -> fmt::
                 res = bt_fmt.frame().symbol(frame, symbol);
             }
         });
-        if stop {
-            return false;
-        }
         #[cfg(target_os = "nto")]
         if libc::__my_thread_exit as *mut libc::c_void == frame.ip() {
             if !hit && start {

--- a/tests/run-make/short-ice/Makefile
+++ b/tests/run-make/short-ice/Makefile
@@ -1,0 +1,9 @@
+include ../tools.mk
+
+# ignore-windows
+
+export RUSTC := $(RUSTC_ORIGINAL)
+export TMPDIR := $(TMPDIR)
+
+all:
+	bash check.sh

--- a/tests/run-make/short-ice/check.sh
+++ b/tests/run-make/short-ice/check.sh
@@ -1,0 +1,36 @@
+#!/bin/sh
+
+RUST_BACKTRACE=1 $RUSTC src/lib.rs -Z treat-err-as-bug=1 1>$TMPDIR/rust-test-1.log 2>&1
+RUST_BACKTRACE=full $RUSTC src/lib.rs -Z treat-err-as-bug=1 1>$TMPDIR/rust-test-2.log 2>&1
+
+short=$(cat $TMPDIR/rust-test-1.log | wc -l)
+full=$(cat $TMPDIR/rust-test-2.log | wc -l)
+rustc_query_count=$(cat $TMPDIR/rust-test-1.log | grep rustc_query_ | wc -l)
+rustc_query_count_full=$(cat $TMPDIR/rust-test-2.log | grep rustc_query_ | wc -l)
+
+begin_count=$(cat $TMPDIR/rust-test-2.log | grep __rust_begin_short_backtrace | wc -l)
+end_count=$(cat $TMPDIR/rust-test-2.log | grep __rust_end_short_backtrace | wc -l)
+
+cat $TMPDIR/rust-test-1.log
+echo "====================="
+cat $TMPDIR/rust-test-2.log
+echo "====================="
+
+echo "short backtrace: $short"
+echo "full  backtrace: $full"
+echo "begin_count: $begin_count"
+echo "end_count  : $end_count"
+echo "rustc_query_count: $rustc_query_count"
+echo "rustc_query_count_full: $rustc_query_count_full"
+
+## backtraces to vary a bit depending on platform and configuration options,
+## here we make sure that the short backtrace of rustc_query is shorter than the full,
+## and marks are in pairs.
+if [ $short -lt $full ] &&
+    [ $begin_count -eq $end_count ] &&
+    [ $(($rustc_query_count + 10)) -lt $rustc_query_count_full ] &&
+    [ $rustc_query_count_full -gt 10 ]; then
+    exit 0
+else
+    exit 1
+fi

--- a/tests/run-make/short-ice/src/lib.rs
+++ b/tests/run-make/short-ice/src/lib.rs
@@ -1,0 +1,7 @@
+fn func(s: &str) {
+    println!("{}", s);
+}
+
+fn main() {
+    func(1);
+}

--- a/tests/ui/panics/default-backtrace-ice.stderr
+++ b/tests/ui/panics/default-backtrace-ice.stderr
@@ -8,6 +8,8 @@ LL | fn main() { missing_ident; }
 stack backtrace:
 (end_short_backtrace)
 (begin_short_backtrace)
+(end_short_backtrace)
+(begin_short_backtrace)
 
 error: the compiler unexpectedly panicked. this is a bug.
 

--- a/tests/ui/panics/short-ice-remove-middle-frames-2.rs
+++ b/tests/ui/panics/short-ice-remove-middle-frames-2.rs
@@ -1,0 +1,61 @@
+// compile-flags:-Cstrip=none
+// run-fail
+// check-run-results
+// exec-env:RUST_BACKTRACE=1
+// ignore-android FIXME #17520
+// ignore-wasm no panic support
+// ignore-openbsd no support for libbacktrace without filename
+// ignore-emscripten no panic
+// ignore-sgx Backtraces not symbolized
+// ignore-fuchsia Backtraces not symbolized
+// ignore-msvc the `__rust_{begin,end}_short_backtrace` symbols aren't reliable.
+
+/// This test case make sure that we can have multiple pairs of `__rust_{begin,end}_short_backtrace`
+
+#[inline(never)]
+fn __rust_begin_short_backtrace<T, F: FnOnce() -> T>(f: F) -> T {
+    let result = f();
+    std::hint::black_box(result)
+}
+
+#[inline(never)]
+fn __rust_end_short_backtrace<T, F: FnOnce() -> T>(f: F) -> T {
+    let result = f();
+    std::hint::black_box(result)
+}
+
+fn first() {
+    __rust_end_short_backtrace(|| second());
+}
+
+fn second() {
+    third(); // won't show up
+}
+
+fn third() {
+    fourth(); // won't show up
+}
+
+fn fourth() {
+    __rust_begin_short_backtrace(|| fifth());
+}
+
+fn fifth() {
+    __rust_end_short_backtrace(|| sixth());
+}
+
+fn sixth() {
+    seven(); // won't show up
+}
+
+fn seven() {
+    __rust_begin_short_backtrace(|| eight());
+}
+
+fn eight() {
+    panic!("debug!!!");
+}
+
+fn main() {
+    first();
+}

--- a/tests/ui/panics/short-ice-remove-middle-frames-2.run.stderr
+++ b/tests/ui/panics/short-ice-remove-middle-frames-2.run.stderr
@@ -1,0 +1,11 @@
+thread 'main' panicked at 'debug!!!', $DIR/short-ice-remove-middle-frames-2.rs:56:5
+stack backtrace:
+   0: std::panicking::begin_panic
+   1: short_ice_remove_middle_frames_2::eight
+   2: short_ice_remove_middle_frames_2::seven::{{closure}}
+   3: short_ice_remove_middle_frames_2::fifth
+   4: short_ice_remove_middle_frames_2::fourth::{{closure}}
+   5: short_ice_remove_middle_frames_2::first
+   6: short_ice_remove_middle_frames_2::main
+   7: core::ops::function::FnOnce::call_once
+note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

--- a/tests/ui/panics/short-ice-remove-middle-frames.rs
+++ b/tests/ui/panics/short-ice-remove-middle-frames.rs
@@ -1,0 +1,57 @@
+// compile-flags:-Cstrip=none
+// run-fail
+// check-run-results
+// exec-env:RUST_BACKTRACE=1
+// ignore-android FIXME #17520
+// ignore-wasm no panic support
+// ignore-openbsd no support for libbacktrace without filename
+// ignore-emscripten no panic
+// ignore-sgx Backtraces not symbolized
+// ignore-fuchsia Backtraces not symbolized
+// ignore-msvc the `__rust_{begin,end}_short_backtrace` symbols aren't reliable.
+
+
+#[inline(never)]
+fn __rust_begin_short_backtrace<T, F: FnOnce() -> T>(f: F) -> T {
+    let result = f();
+    std::hint::black_box(result)
+}
+
+#[inline(never)]
+fn __rust_end_short_backtrace<T, F: FnOnce() -> T>(f: F) -> T {
+    let result = f();
+    std::hint::black_box(result)
+}
+
+fn first() {
+    __rust_end_short_backtrace(|| second());
+    // do not take effect since we already has a inner call of __rust_end_short_backtrace
+}
+
+fn second() {
+    __rust_end_short_backtrace(|| third());
+}
+
+fn third() {
+    fourth(); // won't show up in backtrace
+}
+
+fn fourth() {
+    fifth(); // won't show up in backtrace
+}
+
+fn fifth() {
+    __rust_begin_short_backtrace(|| sixth());
+}
+
+fn sixth() {
+    seven();
+}
+
+fn seven() {
+    panic!("debug!!!");
+}
+
+fn main() {
+    first();
+}

--- a/tests/ui/panics/short-ice-remove-middle-frames.run.stderr
+++ b/tests/ui/panics/short-ice-remove-middle-frames.run.stderr
@@ -1,0 +1,12 @@
+thread 'main' panicked at 'debug!!!', $DIR/short-ice-remove-middle-frames.rs:52:5
+stack backtrace:
+   0: std::panicking::begin_panic
+   1: short_ice_remove_middle_frames::seven
+   2: short_ice_remove_middle_frames::sixth
+   3: short_ice_remove_middle_frames::fifth::{{closure}}
+   4: short_ice_remove_middle_frames::second
+   5: short_ice_remove_middle_frames::first::{{closure}}
+   6: short_ice_remove_middle_frames::first
+   7: short_ice_remove_middle_frames::main
+   8: core::ops::function::FnOnce::call_once
+note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

